### PR TITLE
Impulse_2024

### PR DESCRIPTION
**Pull request by Vignaraj-pai**
# Description

Impulse Module added.

Fixes # (issue)

None

## Dependencies

None

## Type of Change

- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

In process of writing unit tests for urls, views and models.

---

## Checklist

- [x] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: corpus/impulse/forms.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        # set a required attribute based on conditions
+        self.fields["college_name"].required = not self.fields['from_nitk'].widget.value_from_datadict(
+            self.data, self.files, self.add_prefix('from_nitk')
+        )
+        self.fields["roll_no"].required = self.fields['from_nitk'].widget.value_from_datadict(
+            self.data, self.files, self.add_prefix('from_nitk')
+        )
+        self.fields["ieee_membership_no"].required = self.fields['ieee_member'].widget.value_from_datadict(
+            self.data, self.files, self.add_prefix('ieee_member')
+        )
```
Comment (anirudhprabhakaran3) [inline corpus/impulse/forms.py:27]: You can do these checks in the `clean` function in forms. Refer to https://github.com/IEEE-NITK/corpus/blob/main/corpus/embedathon/forms.py#L27

File: corpus/impulse/models.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    member2_name = models.CharField(max_length=200, blank=True, null=True)
+    member2_email = models.EmailField(blank=True, null=True)
+    member2_from_nitk = models.BooleanField(default=True)
+    member2_college_name = models.CharField(
+        max_length=200, default="National Institute of Technology Karnataka"
+    )
+    member2_roll_no = models.CharField(max_length=8, blank=True, null=True)
+    member2_phone = models.CharField(max_length=13, blank=True, null=True)
+    member2_ieee_member = models.BooleanField(default=False)
+    member2_ieee_membership_no = models.BigIntegerField(blank=True, null=True)
```
Comment (anirudhprabhakaran3) [inline corpus/impulse/models.py:27]: What is this `member2` logic? Did not understand this. If your requirement is that every team should have two people or something, have that in the team model (as `leader` and `member`). This way of storing data is not making sense.

File: corpus/impulse/models.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    
+class Team(models.Model):
+    team_name = models.CharField(max_length=200, blank=False, null=False)
+    team_leader = models.ForeignKey(
+        ImpulseUser, on_delete=models.CASCADE, related_name="leader"
+    )
+    payment_status = models.CharField(
+        max_length=1, choices=PAYMENT_STATUS, blank=False, null=False, default="U"
+    )
+    payment_proof = models.ImageField(upload_to="img/impulse/payment_proofs", blank=True, null=True)
```
Comment (anirudhprabhakaran3) [inline corpus/impulse/models.py:40]: Check once that `ImageField` is working for you by using the production Docker Compose. 

cc @nishant-nayak have we setup corpus with media upload capabilities?

File: corpus/impulse/views.py [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+from impulse.forms import ImpulseForm
+from impulse.forms import TeamCreationForm
+from impulse.forms import AnnouncementForm
+from impulse.forms import Member2Form
+from impulse.forms import PaymentProofForm
+
+# from impulse.forms import TeamCreationForm
+
+
+
```
Comment (anirudhprabhakaran3) [inline corpus/impulse/views.py:22]: Why so many blank lines? i think if you remove the comment the pre-commit hooks will edit it properly.

File: corpus/impulse/views.py [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+        messages.error(request, "You have already registered for Impulse!")
+        return redirect("impulse_index")
+    except ImpulseUser.DoesNotExist:
+        pass
+
+    if request.method == "POST":
+        form = ImpulseForm(request.POST)
+        print("Form received")
+        if form.is_valid():
+            print("Form is valid")
```
Comment (anirudhprabhakaran3) [inline corpus/impulse/views.py:170]: Remove `print` statements.

**Review by anirudhprabhakaran3 — CHANGES_REQUESTED** [review]

File: corpus/impulse/forms.py (hunk 1 above)
Comment (Vignaraj-pai) [reply corpus/impulse/forms.py]: Done, implemented clean function.

File: corpus/impulse/models.py (hunk 2 above)
Comment (Vignaraj-pai) [reply corpus/impulse/models.py]: Okay so there can be max of 2 members per team, I thought the invite sending can be overkill just for a 2 team event. 
Yes the member2 logic is weird, my bad. I have put it as member_details (each field listed out) under Team model, is that okay ? or is it better following the invite method ?

File: corpus/impulse/views.py (hunk 4 above)
Comment (Vignaraj-pai) [reply corpus/impulse/views.py]: Removed unnecessary blank lines.

File: corpus/impulse/views.py (hunk 5 above)
Comment (Vignaraj-pai) [reply corpus/impulse/views.py]: Removed all debug statements.

File: corpus/impulse/models.py (hunk 3 above)
Comment (Vignaraj-pai) [reply corpus/impulse/models.py:42]: Hey, sorry but I'm having issues running the production docker_compose, i'm figuring it out but could you @anirudhprabhakaran3 test it out, cause theres a bit of a hurry as they want to begin registrations soon.

**Review by anirudhprabhakaran3 — APPROVED** [review]

**Pull request merged**